app/src/main/java/org/dicio/dicio_android/skills/navigate/NavigateProcessor.java

```diff
@@ -0,0 +1,79 @@
+package org.dicio.dicio_android.skills.navigate;
+
+import static org.dicio.dicio_android.Sentences_en.navigate;
+
+import org.dicio.dicio_android.util.StringUtils;
+import org.dicio.skill.chain.IntermediateProcessor;
+import org.dicio.skill.standard.StandardResult;
+
+import java.util.List;
+
+public class NavigateProcessor extends IntermediateProcessor<StandardResult, NavigateOutput.Data> {
+    @Override
+    public NavigateOutput.Data process(final StandardResult data) throws Exception {
+        final NavigateOutput.Data result = new NavigateOutput.Data();
+
+        if (data == null) {
+            result.failed = true;
+            return result;
+        }
+
+        String placeToNavigate = data.getCapturingGroup(navigate.where);
+
+        if (placeToNavigate == null) {
+            result.failed = true;
+            return result;
+        }
+
+        placeToNavigate = placeToNavigate.trim();
+
+        if (StringUtils.isNullOrEmpty(placeToNavigate)) {
+            result.failed = true;
+            return result;
+        }
+
+        final List<Object> textWithNumbers = ctx().requireNumberParserFormatter()
+                .extractNumbers(data.getCapturingGroup("where")).get();
+
+        // It seems that we detect each digit of the numerical portion of an address as a separate
+        // string, resulting in "9 5 4 6 1 9 avenue" when the speech was actually "9546 19 avenue".
+        // Here we will remove the spaces from between the numbers, so we will end up with
+        // "954619 avenue". This is not ideal, but it seems slightly better than before.
+        // Can we detect addresses better?
```

Comment:
I am not so sure what user input you are referring to here. Wouldn't the user say "nine five four six nineteenth avenue" in that case? So the 19th is separate from the 9546, simply because it is ordinal.

Reply:
Without the number parser formatter, a speech input of "nine" would be transcribed as "nine", whereas we almost always would want that to be transcribed as "9". So that part certainly seems needed. With that out of the way:

For an address of "9546 19 avenue":
- If the speech input is "nine thousand five hundred forty six nineteen avenue" the transcribed output is **correct** as expected "9546 19 avenue"
- If the speech input is "nine thousand five hundred forty six nineteenth avenue" the transcribed output is **wrong** because it seems that the 9546 is multiplied by 1/19 resulting in an output of "502.421052... avenue"
- If the speech input is "nine five four six nineteen avenue" the transcribed output is **wrong** (but close): "9 5 4 6 19 avenue"
- If the speech input is "nine five four six nineteenth avenue" the transcribed output is **wrong**: "9 5 4 0.3157894... avenue". Here we have 6 multiplied by 1/19.
- If the speech input is "ninety five forty six nineteen avenue" the transcribed output is **wrong**: "95 46 19 avenue"

In summary, the problems are:
- Specifying the street number as an ordinal seems to result in some undesired multiplication happening
- Saying the building number using its digits one by one results in undesired spaces in between each one
- Saying the building number using partial grouping of its digits (but not all of them) e.g. "ninety five forty six" also results in undesired spaces in between each partial grouping

I have not attempted to resolve the ordinal multiplication problem at all. I suspect that might require some changes in the number formatter (maybe by allowing the caller to specify whether or not to perform calculations?)

In my code I had tried to resolve the latter 2 problems by removing the spaces in between any numerical strings, however this resulted in also removing the space in between the building number and the street number. For now, I propose removing that space-removal code, meaning we'll still have the original problems but it will work as expected if we specify each of building number and street number as full separate numbers rather than by digit.

Reply:
I proceeded with my proposal of removing the space-removal code. I documented the known issues in a comment beside the relevant code as well. Resolving this conversation, but feel free to re-open.